### ISSUE TITLE
Memory leak issue

### ISSUE THREAD

**Issue by ravifullestop** [bahlo/SwiftGif#6]
I've imported the files and it works just fine with the demo gifs provided but when I load in a 3.9mb 640x400 gif the gif lags and the memory of the device (running in simulator) reaches 223mb, this seems exceedingly high.

Is there a limit for the gif that I'm using that I should take in account?

**Comment by bahlo** [comment]
Hey @ravifullestop, thanks for making an issue! 

I didn't test it with gifs this big, but I will look into it as soon as I have more time. :+1:

**Comment by rhenz** [comment]
+1

**Comment by nleachCHS** [comment]
+1

**Comment by lostpupil** [comment]
+1

**Comment by abdulaziz-jm** [comment]
+1

**Comment by samheather** [comment]
+1 - just loaded a 2.8mb Gif, 2000x2000 and memory usage went to 1GB before crashing out

**Comment by bahlo** [comment]
Sorry I'm only replying so late. :disappointed: 

Since SwiftGif puts all frames in a `UIAnimatedImage` and even duplicates them to match various framerates, the memory leak is easily explained. It's just a lot of memory used.

We could render only a sequence of images at once (like lazy-load, plus drop earlier frames). This would add complexity to the library but help with this issue. 
Another way is to throw an error if the gif is too big, but I don't really like that solution.

What do you guys think?

**Comment by samheather** [comment]
I'm still a little confused at how all the space is been used. Let's take an example:
800x800px GIF with 138 frames and transparency. The GIF is 2.8mb.
With SwiftGif, the memory increase from loading this Gif is ~290mb. So each of those 138 frames is taking 290/138 = 2.1Mb somehow. 
Whereas if I export a frame of my Gif from Photoshop it's about 87kb (24x less). 
Why does extracting each frame from a gif take so much more memory that just rendering each frame individually as a PNG and loading all of those? If we answer that it might help work out a solution?

**Comment by bahlo** [comment]
From the `README.md`:

## How does it work?
Easy, it does the following:

1. Find out the duration of every frame
2. Find the greatest common divisor
3. Add frames accordingly to the greatest common divisor to an array
4. Create an animated UIImage with the frames

----

So if the greatest common divisor is, say, one second, images that are displayed ten seconds would get added ten times. This _may_ be the problem.

**Comment by samheather** [comment]
Yeah it sounds like it. I wonder if there's an optimisation by allowing flexibility in the greatest common divisor, i.e. if we've got frames of 9, 10, 20, 30, 40, we'd have to use GCD of 1, whereas the effect of using 10 would probably be minimal. It could be a toggle-able optimisation - a 10% deviation in the GCD value?

**Comment by petaire92** [comment]
+1

**Comment by petaire92** [comment]
Any news on this ?

**Comment by Mozahler** [comment]
You are using: 
func CGImageSourceCreateWithData(_ data: CFData, _ options: CFDictionary?) -> CGImageSource?
Which Creates an image source that reads from a Core Foundation data object.
returns: An image source. You are responsible for releasing this object using CFRelease.

I don't see any calls to CFRelease in your code. This means that the memory used during each call is retained and never deallocated.

**Comment by tommylo** [comment]
Even the view controller is deinit, the UIImage with gif is not released.

**Comment by daemelchi** [comment]
https://github.com/daemelchi/SwiftGif/commit/1d8f002ba8cc596f3262b25cc61ad33624896140

**Comment by FouZ** [comment]
The fix proposed by @daemelchi should be pull requested. The defer fixed the leak I had in my app.

**Comment by miktav** [comment]
The fix proposed by @daemelchi seems to have resolved the leak. I also had to set the UIImage property to nil in viewDidDisappear.

**Comment by alessandro-martin** [comment]
@bahlo Could you please merge @daemelchi's fix and issue a new release?

**Comment by bahlo** [comment]
Hey guys, sorry for the problems. @daemelchi can you please create a PR so we can resolve this?

**Comment by SolDiez90** [comment]
Switching view controllers doesn't release the gif from memory which causes massive memory leaks especially for quite big gifs. I solved this by loading the UIImage in an appdelegate class but this is just a workaround.

**Comment by JayeshPoriya** [comment]
1) When I switching view controller, Gif doesn't release memory. How can I deal with this issue? 
How can I manage OR solve this issue? 
2) When I try to load some other GIF with same viewcontroller on same imageview (On different button action), It's going to very horrible memory leak. It's going to 650MB. 
Please Help.
